### PR TITLE
Continuous underline dotted

### LINKED ISSUE #4349: Take column into account when rendering cells for perfect patterned underline rendering
Forking off from https://github.com/xtermjs/xterm.js/issues/4060

> Thinking about this and the fact that I want xterm.js rendering to be as great as possible, I think we should support multiple glyphs depending on the x-axis when necessary. For the dotted line specifically, say we had a dot of size `window.devicePixelRatio` with the same sized gap in-between, so there would be at most 2 possible variants for every glyph:
> 
> ```
> Even:
> |* * * * |
> 
> Odd:
> |* * * * *|
> | * * * * |
> ```
> 
> To store this as a unique glyph we need to make the glyph key different, this could be done by asking what underline variant it is and then store it as a number, maybe capped at 4 (2 bits) or 8 variants (3 bits)?
> 
> This would allow us make a non-repeated curvy line and not restrict us to a single cell for both the up and down parts of the wave, if that ends up looking better of course.

Current:

image
image

Desired:

image
image

### PR DESCRIPTION
**Pull request by tisilent**
fix #4349 
trial version
Font size above 26 is unstable.
Only processed webgl.
@Tyriar Is this idea feasible?

### CODE REVIEW

**Comment by tisilent** [comment]
emmm, file format ...... 😭

**Comment by tisilent** [comment]
企业微信截图_16923578691709
企业微信截图_16923578691709
This should be a reasonable situation.
size 21  ,   22

File: addons/xterm-addon-webgl/src/WebglRenderer.ts [hunk 1]
```diff
@@ -481,13 +500,28 @@ export class WebglRenderer extends Disposable implements IRenderer {
...
-            this._model.cells[i + RENDER_MODEL_FG_OFFSET] === this._cellColorResolver.result.fg &&
-            this._model.cells[i + RENDER_MODEL_EXT_OFFSET] === this._cellColorResolver.result.ext) {
+          this._model.cells[i + RENDER_MODEL_BG_OFFSET] === this._cellColorResolver.result.bg &&
+          this._model.cells[i + RENDER_MODEL_FG_OFFSET] === this._cellColorResolver.result.fg &&
+          this._model.cells[i + RENDER_MODEL_EXT_OFFSET] === this._cellColorResolver.result.ext &&
+          this._model.cells[i + RENDER_MODEL_VARIANT_OFFSET] === variantOffset
+        ) {
+          if (cell.extended.underlineStyle === UnderlineStyle.DOTTED) {
+            const s = this._charSizeService.width;
+            variantOffset = ((deviceCellWidth! * chWidth) - variantOffset) % (lineWidth * 2);
```
Comment (Tyriar) [inline addons/xterm-addon-webgl/src/WebglRenderer.ts:519]: For dotted, I would expect there only to be _up to_ `2 * dot_size` variants. So when the dot size is 1px, these would be the 2 variants for `s`:

image

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: addons/xterm-addon-webgl/src/WebglRenderer.ts (hunk 1 above)
Comment (tisilent) [reply addons/xterm-addon-webgl/src/WebglRenderer.ts:519]: Yes, I also think so. The formula for calculating offset has not been written yet.

File: addons/xterm-addon-webgl/src/WebglRenderer.ts [hunk 2]
```diff
@@ -481,13 +500,28 @@ export class WebglRenderer extends Disposable implements IRenderer {
...
 
         // Nothing has changed, no updates needed
         if (this._model.cells[i] === code &&
-            this._model.cells[i + RENDER_MODEL_BG_OFFSET] === this._cellColorResolver.result.bg &&
-            this._model.cells[i + RENDER_MODEL_FG_OFFSET] === this._cellColorResolver.result.fg &&
-            this._model.cells[i + RENDER_MODEL_EXT_OFFSET] === this._cellColorResolver.result.ext) {
+          this._model.cells[i + RENDER_MODEL_BG_OFFSET] === this._cellColorResolver.result.bg &&
+          this._model.cells[i + RENDER_MODEL_FG_OFFSET] === this._cellColorResolver.result.fg &&
+          this._model.cells[i + RENDER_MODEL_EXT_OFFSET] === this._cellColorResolver.result.ext &&
+          this._model.cells[i + RENDER_MODEL_VARIANT_OFFSET] === variantOffset
```
[the remark this answers is not in the document]
Comment (tisilent) [reply addons/xterm-addon-webgl/src/WebglRenderer.ts:515]: 😨   I've sorted out the code and it's a bit messy. I've already figured out how to solve 2px+.

**Comment by tisilent** [comment]
??...... why.

**Comment by tisilent** [comment]
close,Resubmit. :sob:.

**Pull request closed without merging**